intrahost.py

```diff
@@ -129,157 +169,211 @@ def parser_tabfile_rename(parser=argparse.ArgumentParser()):
-                    for a,n in rows[s]:
-                        f = float(n)/tot_n
-                        if a.startswith('I'):
-                            # insertion allele is first ref base, plus inserted bases, plus subsequent ref bases
-                            a = consAllele[0] + a[1:] + consAllele[1:]
-                        elif a.startswith('D'):
-                            # deletion is the first ref base, plus remaining ref seq with the first few positions dropped off
-                            a = consAllele[0] + consAllele[1+int(a[1:]):]
-                        elif a in ('i','d'):
-                            # this is vphaser's way of saying the "reference" (majority/consensus) allele, in the face of other indel variants
-                            a = consAllele
+        # one reference chrom at a time
+        with open(refFasta, 'rU') as inf:
+            for ref_sequence in Bio.SeqIO.parse(inf, 'fasta'):
+
+                # read in all iSNVs for this chrom and map to reference coords
+                data = []
+                for s in samples:
+                    for row in util.file.read_tabfile(samp_to_isnv[s]):
+                        s_chrom = samp_to_cmap[s].mapBtoA(ref_sequence.id)
+                        if row[0] == s_chrom:
+                            row = {'sample':s, 'CHROM':ref_sequence.id,
+                                's_chrom':s_chrom, 's_pos':int(row[1]),
+                                's_alt':row[2], 's_ref': row[3],
+                                'n_libs':row[7], 'lib_bias': row[8],
+                                'alleles':list(x.split(':') for x in row[9:] if x)}
+                            # make a sorted allele list
+                            row['allele_counts'] = list(sorted(
+                                [(a,int(f)+int(r)) for a,f,r in row['alleles']],
+                                key=(lambda x:x[1]), reverse=True))
+                            # reposition vphaser deletions minus one to be consistent with
+                            # VCF conventions
+                            if row['s_alt'].startswith('D'):
+                                for a,n in row['allele_counts']:
+                                    assert a[0] in ('D','i')
+                                row['s_pos'] = row['s_pos']-1
+                            # map position back to reference coordinates
+                            row['POS'] = samp_to_cmap[s].mapAtoB(s_chrom, row['s_pos'], side = -1)[1]
+                            row['END'] = samp_to_cmap[s].mapAtoB(s_chrom, row['s_pos'], side = 1)[1]
+                            data.append(row)
+            
+                # sort all iSNVs (across all samples) and group by position
+                data = sorted(data, key=(lambda row: row['POS']))
+                data = itertools.groupby(data, lambda row: row['POS'])
+            
+                # process one reference position at a time
+                for pos, rows in data:
```

Comment:
So, in practice, very few, if any, callers will correctly handle this, and also, the GATK IndelRealigner (which we call prior to V-Phaser) should try to minimize the real-life scenarios in which we might encounter something like this.  Currently, this code does not attempt to do anything with this, and they are reported on independently.

However, one could conceivably replace the itertools.groupby line with something that does some sort of overlapping intervals function on the POS-END interval (perhaps using an R-Tree for optimal implementation). That would solve everything actually, the rest of the downstream code should merge everything smartly.

I'm not sure how much to worry about this scenario though. It wouldn't be hard to detect the scenario and push a log.warn if it happens.  After the groupby, everything is already sorted on POS.  After the code block that determines "end" based on the longest variant, we could save it in some variable that persists outside of the for loop and then at the top of the loop, if max_previous_end >= pos: log.warn("overlapping variants in VCF!").